src/NuGet.Core/NuGet.ProjectModel/LockFile/Utf8JsonStreamLockFileItemConverter.cs

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Text.Json;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// A <see cref="Utf8JsonStreamReaderConverter{T}"/> to allow read JSON into <see cref="LockFileItem"/>
+    /// </summary>
+    internal class Utf8JsonStreamLockFileItemConverter<T> : IUtf8JsonStreamReaderConverter<T> where T : LockFileItem
+    {
+        public T Read(ref Utf8JsonStreamReader reader)
+        {
+            var genericType = typeof(T);
+
+            if (reader.TokenType != JsonTokenType.PropertyName)
+            {
+                throw new JsonException("Expected PropertyName, found " + reader.TokenType);
+            }
+
+            //We want to read the property name right away
+            var lockItemPath = reader.GetString();
+            LockFileItem lockFileItem;
+            if (genericType == typeof(LockFileContentFile))
+            {
+                lockFileItem = new LockFileContentFile(lockItemPath);
```

Comment:
I think this will result in a lot less branches during parsing if you make this a `Func<string, T> lockFileItemCreator` that gets passed into the constructor instead of branching based on the generic type.